### PR TITLE
.rscignore update for .gitignore functionality

### PR DESCRIPTION
**Pull request by BrianLang**
# Hierarchical .rscignore Implementation

## Summary

This PR implements **hierarchical .gitignore-style behavior** for `.rscignore` files as discussed in #1178, making them work like `.gitignore` files where parent directory patterns affect subdirectories and allowing general wildcard matching as in `.gitignore` files.

### Core Behavior Change
- **Before**: `.rscignore` patterns only affected files in the same directory and use literal string matching to select ignored files.
- **After**: `.rscignore` patterns inherit to subdirectories and support most .gitignore supported patterns (see below)

```bash
# Example: Root .rscignore with "*.log"
project/
├── .rscignore          # "*.log"
├── app.log             # ✅ Ignored (same as before)
├── src/
│   ├── debug.log       # ✅ NOW ignored (new behavior)
│   └── main.R          # ✅ Preserved
└── tests/
    └── error.log       # ✅ NOW ignored (new behavior)
```

### Basic Patterns
- *.log (wildcard patterns)
- temp* (prefix patterns)
- cache/ (directory patterns)
- /README (root-relative patterns)

### Hierarchical Behavior
- Parent .rscignore affects subdirectories
- Child .rscignore overrides parent patterns
- !pattern (negation patterns work across directories)

### Advanced Patterns
- **/cache.tmp (matches at any depth)
- **/logs/*.log (complex globstar patterns)
- \# comments and empty lines
- Pattern precedence (most specific wins)

### Examples That Work
```
# Root .rscignore
*.log
!important.log
cache/
**/node_modules/

# src/.rscignore  
!debug.log
temp*
```

### Advanced Pattern Syntax
- [abc] (character classes)
- [a-z] (character ranges)
- *.{js,css} (brace expansion)
- file\*.txt (escape sequences)

### New Functions Added
- `collectHierarchicalPatterns()` - Collects patterns from current directory up to project root
- `applyRscignorePatterns()` - Enhanced to use hierarchical patterns with fallback
- `applyRscignorePatternsLegacy()` - Preserves old behavior during transition

### Modified Functions
- `bundleFiles()` - Now passes bundle root context
- `ignoreBundleFiles()` - Uses hierarchical pattern collection
- All recursive file functions - Thread bundle root through call chain

### Pattern Precedence Rules
- **Child overrides parent**: Subdirectory `.rscignore` patterns take precedence
- **Negation support**: `!pattern` in child can un-ignore parent's `pattern`
- **Standard .gitignore behavior**: Follows established gitignore precedence rules

### Breaking Change Strategy
Following RStudio/Posit patterns for managed breaking changes:

```r
# Default: New hierarchical behavior
options(rsconnect.rscignore.legacy = FALSE)  # Default

# Temporary: Legacy behavior with deprecation warning
options(rsconnect.rscignore.legacy = TRUE)   # Deprecated
```

## Test Coverage
- Hierarchical inheritance behavior
- Pattern precedence rules (child overrides parent)
- Negation patterns across directories
- Gitignore-style patterns (wildcards, directory patterns)
- Error handling and graceful fallback
- Legacy vs hierarchical behavioral differences
- Edge cases and parameter validation

### Core Implementation
- `R/bundleFiles.R` - Main hierarchical logic and lifecycle management
- `man/*.Rd` - Documentation for new exported functions

### Test Suite  
- `tests/testthat/test-hierarchical-rscignore.R` - Comprehensive hierarchical behavior tests
- `tests/testthat/_snaps/bundleFiles.md` - Updated snapshots for new behavior

### Documentation
- `NEWS.md` - Suggestion for breaking change announcement and migration guide

### CODE REVIEW

File: R/bundleFiles.R [hunk 1]
```diff
@@ -284,3 +287,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+  
+  # Handle NULL case
+  if (is.null(legacy_mode)) {
+    legacy_mode <- FALSE
+  }
+  
+  if (legacy_mode) {
+    # Issue deprecation warning for legacy behavior
+    lifecycle::deprecate_warn(
+      when = "1.6.0",
```
Comment (BrianLang) [inline R/bundleFiles.R]: unclear when this should be.

File: R/bundleFiles.R [hunk 2]
```diff
@@ -284,3 +287,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+  if (legacy_mode) {
+    # Issue deprecation warning for legacy behavior
+    lifecycle::deprecate_warn(
+      when = "1.6.0", 
+      what = I("Legacy .rscignore behavior"),
+      details = c(
+        "Directory-scoped .rscignore patterns are deprecated.",
+        "Hierarchical .gitignore-style behavior is now the default.",
+        "Update your .rscignore files to use '!' negation patterns if needed.",
+        "Set `options(rsconnect.rscignore.legacy = FALSE)` to remove this warning."
```
Comment (BrianLang) [inline R/bundleFiles.R:330]: Happy for feedback on what this message would say

File: tests/testthat/test-rscignore-gitignore.R
Comment (BrianLang) [inline tests/testthat/test-rscignore-gitignore.R:1]: happy to integrate this (include these) into the test-bundleFiles.R to be consistent with normal approach.

File: R/bundleFiles.R [hunk 3]
```diff
@@ -127,14 +132,15 @@ recursiveBundleFiles <- function(
...
+  ignoreFiles = TRUE,
+  bundle_root = rootDir
 ) {
   # generate a list of files at this level
   if (is.null(contents)) {
     contents <- list.files(dir, all.files = TRUE, no.. = TRUE)
   }
   if (ignoreFiles) {
-    contents <- ignoreBundleFiles(dir, contents)
+    contents <- ignoreBundleFiles(dir, contents, bundle_root)
```
Comment (karawoo) [inline R/bundleFiles.R:143]: Do we need the new `bundle_root` argument to this function or could we pass the existing `rootDir` argument to `ignoreBundleFiles` instead?

File: R/bundleFiles.R [hunk 4]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+    }
+    legacy_mode <- converted
+  }
+
+  # Handle NULL case
+  if (is.null(legacy_mode)) {
+    legacy_mode <- FALSE
+  }
+
+  if (legacy_mode) {
```
Comment (karawoo) [inline R/bundleFiles.R:321]:
```suggestion
  if (isTRUE(legacy_mode)) {
```

this way would be just a little simpler

File: R/bundleFiles.R (hunk 2 above)
Comment (karawoo) [reply R/bundleFiles.R:330]: Since people will only see this if they've explicitly opted into the legacy behavior, I think we don't need to tell them what option to use to silence it. Could we instead include a link to the documentation that describes the change and how to update your .rscignore files?

File: R/bundleFiles.R [hunk 5]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+      )
+    )
+    # Use directory-scoped behavior
+    return(applyDirectoryScopedPatterns(contents, dir))
+  }
+
+  # NEW DEFAULT: Use hierarchical behavior
+  if (is.null(bundle_root)) {
+    bundle_root <- dir  # Fallback for backward compatibility
+  }
```
Comment (karawoo) [inline R/bundleFiles.R:340]: Is there any reason not to have the `bundle_root` argument default to `dir` instead of setting it here?

File: R/bundleFiles.R [hunk 6]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+    contents <- setdiff(contents, ".rscignore")
+    return(contents)
+
+  }, error = function(e) {
+    # Fallback to directory-scoped behavior on error
+    warning("Error in hierarchical pattern processing: ", e$message, call. = FALSE)
+    warning("Falling back to directory-scoped patterns", call. = FALSE)
+    return(applyDirectoryScopedPatterns(contents, dir))
+  })
+}
```
Comment (karawoo) [inline R/bundleFiles.R:358]: I think this `tryCatch()` has the potential to obfuscate the behavior a little too much. I think we should go ahead and fail fast if an error occurs, and the error message can point people to the `rsconnect.rscignore.legacy` option.

File: R/bundleFiles.R [hunk 7]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+
+#' Directory-scoped .rscignore pattern application
+#'
+#' Applies .rscignore patterns only from the current directory,
+#' without hierarchical inheritance from parent directories.
+#'
+#' @param contents File contents to filter
+#' @param dir Directory to check for .rscignore
+#' @return Filtered contents
+applyDirectoryScopedPatterns <- function(contents, dir) {
```
Comment (karawoo) [inline R/bundleFiles.R:368]: Am I correct in understanding that this function replaces this logic that was originally in `ignoreBundleFiles`:

```r
  if (".rscignore" %in% contents) {
    ignoreContents <- readLines(file.path(dir, ".rscignore"))
    contents <- setdiff(contents, c(ignoreContents, ".rscignore"))
  }

  contents
```

If so, I'd suggest we reuse that logic exactly to make sure that we don't inadvertently change the legacy behavior (though I appreciate the thought you put into error handling here).

File: R/bundleFiles.R [hunk 8]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+      if (!is.null(pattern_obj)) {
+        patterns <- append(patterns, list(pattern_obj))
+      }
+    }
+
+    return(patterns)
+  }, error = function(e) {
+    warning("Error reading .rscignore file: ", e$message)
+    return(list())
+  })
```
Comment (karawoo) [inline R/bundleFiles.R:495]: Can we remove this `tryCatch` and raise an informative error if the file can't be parsed?

File: R/bundleFiles.R [hunk 9]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+  } else if (line == "**/") {
+    # **/ matches all directories
+    dir_only <- TRUE
+    line <- "*"
+  }
+
+  # Issue warning if needed
+  if (!is.null(warning_msg)) {
+    warning(warning_msg)
+  }
```
Comment (karawoo) [inline R/bundleFiles.R:557]: Let's skip the warning here and simplify to: 

```suggestion
  if (grepl("\\*{2,}$", line)) {
    # two or more asterisks alone match everything
    line <- "*"
  } else if (line == "**/") {
    # **/ matches all directories
    dir_only <- TRUE
    line <- "*"
  }
```

File: R/bundleFiles.R [hunk 10]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+    # Move up one level
+    parent_dir <- dirname(search_dir)
+
+    # Stop if we can't go higher (filesystem root)
+    if (parent_dir == search_dir) {
+      break
+    }
+
+    search_dir <- parent_dir
+  }
```
Comment (karawoo) [inline R/bundleFiles.R:450]: Do we need to walk up the file hierarchy or could we simply split the path on `.Platform$file.sep` and retain all directories between `bundle_root` and `current_dir`? As a bonus, then we might not need to reverse `directories` down at line 455

File: R/bundleFiles.R [hunk 11]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+    warning("Error reading .rscignore file: ", e$message)
+    return(list())
+  })
+}
+
+#' Parse a single pattern line
+#'
+#' @param line Raw line from .rscignore file
+#' @return Pattern object or NULL if line should be skipped
+parseSinglePattern <- function(line) {
```
Comment (karawoo) [inline R/bundleFiles.R:502]: This is all super clear, thank you 🙏

File: R/bundleFiles.R [hunk 12]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+    } else {
+      target <- basename(file_path)  # Just the filename
+    }
+
+    # Perform match
+    grepl(regex_pattern, target)
+  }, error = function(e) {
+    warning("Pattern matching error for: ", pattern$raw, " - ", e$message)
+    FALSE
+  })
```
Comment (karawoo) [inline R/bundleFiles.R:617]: Here again, if we can't parse then I'd rather raise an informative error rather than succeed with a warning.

File: R/bundleFiles.R [hunk 13]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+  }
+
+  # Track which files are ignored
+  file_status <- setNames(rep(FALSE, length(file_list)), file_list)
+
+  # Process patterns in order (parent to child, within-file order preserved)
+  # Later patterns override earlier patterns
+  for (pattern in patterns) {
+    for (file in file_list) {
+      # Hierarchical pattern matching logic (formerly in matchPatternHierarchical)
```
Comment (karawoo) [inline R/bundleFiles.R:664]:
```suggestion
```

seems like this references a previous implementation that's not relevant anymore

File: R/bundleFiles.R [hunk 14]
```diff
@@ -284,3 +288,423 @@ detectLongNames <- function(bundleDir, lengthLimit = 32) {
...
+        # Normalize paths for comparison
+        pattern_dir <- normalizePath(pattern_dir, mustWork = FALSE)
+        file_dir <- normalizePath(file_dir, mustWork = FALSE)
+
+        # If directories match, compare just the filename
+        if (pattern_dir == file_dir) {
+          file_basename <- basename(file)
+          matches <- matchPattern(file_basename, pattern, current_dir)
+        }
+      } else {
```
Comment (karawoo) [inline R/bundleFiles.R:689]: Is there any reason not to make this logic part of `matchPattern`?

File: tests/testthat/test-rscignore-gitignore.R
Comment (karawoo) [reply tests/testthat/test-rscignore-gitignore.R:1]: Yes, that would be great, thanks!

File: tests/testthat/test-rscignore-gitignore.R [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+
+  # Grandchild: re-ignore temp.txt in src/utils/
+  writeLines("temp.txt", file.path(dir, "src/utils/.rscignore"))
+
+  result <- bundleFiles(dir)
+
+  expect_false("temp.txt" %in% result)           # Root: ignored
+  expect_true("src/temp.txt" %in% result)        # Child: un-ignored (overrides parent)
+  expect_false("src/utils/temp.txt" %in% result) # Grandchild: re-ignored (overrides parent negation)
+  expect_true("src/main.R" %in% result)          # Unaffected
```
Comment (karawoo) [inline tests/testthat/test-rscignore-gitignore.R:178]: awesome 🎉

File: R/bundleFiles.R [hunk 16]
```diff
@@ -13,10 +13,12 @@
...
-#' *  You can exclude additional files by listing them in in a `.rscignore`
-#'    file. This file must have one file or directory per line (with path
-#'    relative to the current directory). It doesn't support wildcards, or
-#'    ignoring files in subdirectories.
+#' *  You can exclude additional files by listing them in a `.rscignore`
+#'    file. This file follows .gitignore-style syntax with one pattern per line.
+#'    Patterns support wildcards (* and ?), directory patterns (dir/), and
+#'    negation (!pattern). Character classes ([abc], [0-9]) and brace expansion
+#'    ({a,b,c}, {1..3}) are not currently supported. Patterns in parent
+#'    directories affect subdirectories hierarchically.
```
Comment (karawoo) [inline R/bundleFiles.R:21]:
```suggestion
#' *  You can exclude additional files by listing them in a `.rscignore`
#'    file. This file follows .gitignore-style syntax with one pattern per line.
#'    Patterns support wildcards (* and ?), directory patterns (dir/), and
#'    negation (!pattern). Character classes (`[abc]`, `[0-9]`) and brace
#'    expansion (`{a,b,c}`, `{1..3}`) are not currently supported. Patterns in
#'    parent directories affect subdirectories hierarchically.
```

The brackets here were causing r cmd check complaints, which should be fixed by code formatting them.

```

❯ checking Rd cross-references ... WARNING
  Missing link(s) in Rd file 'listDeploymentFiles.Rd':
    ‘abc’ ‘0-9’
  
  See section 'Cross-references' in the 'Writing R Extensions' manual.
  
  Found the following Rd file(s) with Rd \link{} targets missing package
  anchors:
    listDeploymentFiles.Rd: abc, 0-9
  Please provide package anchors for all Rd \link{} targets not in the
  package itself and the base packages.

❯ checking Rd files ... NOTE
  checkRd: (-1) listDeploymentFiles.Rd:47: Lost braces; missing escapes or markup?
      47 | ({a,b,c}, {1..3}) are not currently supported. Patterns in parent
         |  ^
  checkRd: (-1) listDeploymentFiles.Rd:47: Lost braces; missing escapes or markup?
      47 | ({a,b,c}, {1..3}) are not currently supported. Patterns in parent
         |           ^

```

**Review by karawoo — COMMENTED** [review]
Hi @BrianLang, thank you so much for taking this on! This feels really on the right track and I appreciate the effort and the thorough test coverage. 

I know this PR is a draft still, but I hope you don't mind me chiming in early with some comments. Feel free to disregard for now if you're still working on things, and you can tag me back in whenever you're ready for another look.

**Comment by BrianLang** [comment]
Thanks so much for the thorough review, i appreciate the time you took, I will address them in the coming days.
